lib/tasks/config.rake

```diff
@@ -181,4 +181,10 @@ Rake::Task['db:seed'].enhance do
       Setting.create! setting.update(:category => "Setting::General")
     end
   end
+
+  if Rails.env.production?
+    File.open('/var/lib/foreman/database_seeded') do |file|
```

Comment:
To be clear, I think we should *never* hardcode a path. We already have code that makes sure `/var/lib/foreman/db` is a symlink to `/usr/share/foreman/db`. That means I think we should use `File.join(Rails.root, 'db', 'seeded')` because it'll work in both production and development. Note 100% sure about permissions though.

https://github.com/theforeman/foreman-packaging/blob/51812c4a35e4b0e010022777371a80d8939921db/packages/foreman/foreman/foreman.spec#L823

Reply:
I think we should _never_ save any state in files. This should be saved in the database.

Reply:
Do you have any suggestions on how to solve this given:

 1) Puppet module doesn't want to run db:seed everytime
 2) RPMs shouldn't rely on the needing the database to be running to get updated

Reply:
Currently we have a setting that it was seeded, but we have no way to indicate a seed is needed again.

With database migrations, we store all the files we ran but seeds should be idempotent and files are ran again. That means you need to store the time. You could use `max(seed.mtime for seed in seeds)` and store that. It probably means that when you add a new node to the cluster you'll seed again.

In these cases I always look at what Django has done and they don't have seeds but rather data migrations which are handled in the same way as schema migrations.

Taking a step back, we could treat seeds the same way since they're in `seeds.d`. You never modify a seed but only add files. If you need to alter, you can rename it. However, that might be diverting from vanilla rails too much. Some seeds also depend on the order.

Reply:
> 1. Puppet module doesn't want to run db:seed everytime

Basically, I'd extend `db/seeds.rb` to save the following information in a database table: `rails_engine_name`, `seed_filename`, `seed_file_sha256` and `migration_date`. If the sha256 changes or the entry does not exist, we need to run the seed.

> 2\. RPMs shouldn't rely on the needing the database to be running to get updated

I'd just run the seed logic every time the rails app starts. That needs the database running anyways. Neither the rpm, the admin or puppet needs to do anything here.

Reply:
> > 1. Puppet module doesn't want to run db:seed everytime
> 
> Basically, I'd extend `db/seeds.rb` to save the following information in a database table: `rails_engine_name`, `seed_filename`, `seed_file_sha256` and `migration_date`. If the sha256 changes or the entry does not exist, we need to run the seed.

I like the idea, but would it also work with cases like:

https://github.com/theforeman/foreman/blob/develop/db/seeds.d/020-permissions_list.rb
https://github.com/theforeman/foreman/blob/a9b03964bd5cb3a0f039af8b6f97a4d8686c289c/db/seeds.d/020-roles_list.rb#L1
https://github.com/theforeman/foreman/blob/a9b03964bd5cb3a0f039af8b6f97a4d8686c289c/db/seeds.d/030-permissions.rb#L1-L5

I believe this is done so plugins can extend the permissions list. This probably works if you're very strict that seeds can only require other seed files.

> > 2. RPMs shouldn't rely on the needing the database to be running to get updated
> 
> I'd just run the seed logic every time the rails app starts. That needs the database running anyways. Neither the rpm, the admin or puppet needs to do anything here.

This feels like overkill. However, if you you can determine if you need to run by only querying a single table and `map(sha256, seed.d/*)` it may be cheap enough.

Odd edge case, but calculating the bcrypt cost might be misleading if the app is starting up because it'll likely saturate the CPU core with unrelated work. https://github.com/theforeman/foreman/blob/a9b03964bd5cb3a0f039af8b6f97a4d8686c289c/db/seeds.d/010-calibrate_bcrypt_cost.rb#L1-L2

I'd like to see this encapsulated and exposed in a rake task as well.

What do you think about trucating the seed table to make sure you purge all old, possibly non-existing seed entries when seeding?

Reply:
> I believe this is done so plugins can extend the permissions list. This probably works if you're very strict that seeds can only require other seed files.

I believe there is a proper plugin API for this. Templates might be an issue, but we can find a solution for this. Maybe generating the sha over the templates as well.

> This feels like overkill. However, if you you can determine if you need to run by only querying a single table and map(sha256, seed.d/*) it may be cheap enough.

Sure, that should work as well.

> What do you think about trucating the seed table to make sure you purge all old, possibly non-existing seed entries when seeding?

We could remove all entries where the seed files do not exist anymore.

Reply:
Sounds a bit like we are implementing database migration tracking instead of just using database migrations in Rails. Is there enough of a difference to support both?

Reply:
Of course you can use the simple timestamp approach and just store that value in a (hidden) setting. If max(filemtimes) > the saved timestamp, just run the seeds.

Reply:
Either of y'all mind doing a recap of where we landed with the implementation idea?

Reply:
I'm not sure if we actually reached a conclusion.
@ehelms: Do you mind describing your initial idea? I'm still not sure what you had initially in mind.

Reply:
The initial idea was:

 1) When database seeding occurs, create a file on disk
 2) puppet module will run db:seed if that file does not exist
 3) RPM packages or otherwise would remove this file on update to trigger a database seed by the puppet module

This prevents the RPMs from touching the database and the puppet module from running database seed everytime and triggering a false positive change event within puppet.

I agree it's not the nicest way but it achieved the end result. There is arguably a larger problem in how we overuse database seeding. Their intent as I understand is to load initial data into a database and in some cases we apply this pattern to upgrades. If we had a hard stance on initial data we'd run seeding just once and never again.

Reply:
The more I think about it is, that we should run the seeds on the application start. That would force us to keep seeds fast and figure out a way to run them just once.

Reply:
I wonder if that is a change we could enforce as 2.0? Any edge cases would then force us to cover them either through a plugin API or use of database migrations. @tbrisker for your thoughts

Reply:
It feels to me like using a file as a signal between rpm and puppet is not the best solution for this, for reasons mentioned before. 
Perhaps, we could calculate a hash that includes all the seed files that ran in https://github.com/theforeman/foreman/blob/develop/db/seeds.rb#L26 (where we already have the plugin seeds also) and save that in the database, than at startup calculate the hash again and if it doesn't match automatically rerun the seeds. This will catch any changes to any seed or configuration, and actually remove the "rake db:seed" step from the installation (as it will automatically execute on the first startup since the initial hash will be blank). This will also cover changes to e.g. the permissions list.
Regarding using seeds to modify data in existing installation - I agree that should be done in migrations generally. However, I think in some cases a migration won't work for new installations since some data won't be seeded yet so it won't be modified (e.g. assigning resources to the default taxonomies). I'm not sure how we can enforce making sure we have a migration that matches any change in the seed though, other than by the reviewers asking for it.